aerogear/data-sync-server
# added security and validation to apollo server

added sane defaults and the ability to configure the values.

## Motivation
https://issues.jboss.org/browse/AEROGEAR-7721
https://issues.jboss.org/browse/AEROGEAR-7719

## What
Added some security features to the sync server to validate queries based on depth, complexity and also added a configurable option to allow or disallow introspection.

## Why
Depth limiting lowers the risk of cyclical requests which could be used as a DOS attack. Complexity ensures the server runs at a healthy level. The ability to disable Introspection means increased security.

## How
Using: 
https://github.com/4Catalyzer/graphql-validation-complexity
https://github.com/stems/graphql-depth-limit

## Verification Steps
1. Run the server from this branch. 
1. In the playground create multiple queries and run them so that there are no errors.
1. In the .env file, set INTROSPECTION to false.
1. In the playground you should now see errors as playground uses introspection to gather schema information.
1. Reset the INTROSPECTION env var.
1. Set the DEPTH_LIMIT env var to 1. This should mean all queries are denied as any returned data has a depth of at least 1.
1. Increase the depth gradually and verify that the results of the queries match what is expected.
1. Repeat the above 3 steps for complexity, setting it to a low value to verify it works.

## Progress
- [x] Add validation to server
- [x] Add introspection
- [x] Add sane defaults
- [x] Allow values to be easily configurable
- [ ] Add cost directives
